### PR TITLE
Add support for TwoPaneView and Surface Duo

### PR DESCRIPTION
**Pull request by jeromelaban**
## PR Type

What kind of change does this PR introduce?
- Feature

## What is the new behavior?

- Add support for [TwoPaneView](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.twopaneview?view=winui-2.3) control.
- Add support for `ApplicationView.GetSpanningRects`
- Add base support for API Extensibility through `Uno.Foundation.Extensibility.ApiExtensibility` and `ApiExtensionAttribute`
- Add support for Surface Duo through the `Uno.UI.DualScreen` package
- Fix invalid conversion when using ThemeResource (e.g. Color resource to Brush property)

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current [supported SDKs](https://github.com/unoplatform/uno/blob/master/README.md#uno-features)
- [x] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [x] [Wasm UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Updated the [Release Notes](https://github.com/unoplatform/uno/tree/master/doc/ReleaseNotes)
- [ ] Associated with an issue (GitHub or internal)

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: doc/articles/uno-development/api-extensions.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# API Extensions
```
Comment (jeromelaban) [inline doc/articles/uno-development/api-extensions.md]: I'm particularly interested in comments on this document

File: doc/articles/uno-development/api-extensions.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+Uno provides a simple mecanism which allows for external provides to provide an implementation for some known interfaces. The goal behind this is two fold:
+
+- Provide a way to remove some types of dependencies from the main Uno.UI package, reducing the payload size of an app if some features are not required
+- Provide a way for developers to provide a custom implementation of a known WinRT or WinUI api for which the default implementation is not provided
+
+A common scenario can be found on Android, where adding an external dependency can increase the build time and payload size unnecessarily.
+
+## Declaring an extension
+
+In a nuget package, depending on the Uno.UI package, define the follow code:
```
Comment (davidjohnoliver) [inline doc/articles/uno-development/api-extensions.md]:
```suggestion
In a library, referencing the Uno.UI package, define the following code:
```

File: doc/articles/uno-development/api-extensions.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+public class MySomeExtensibleType : ISomeExtensibleType
+{
+    public MySomeExtensibleType(object owner)
+    {
+    
+    }
+}
+```
+
+When a nuget package containing such a declaration is found in the currently built application, the App.InitializeComponent() method will automatically add the following code to the app startup:
```
Comment (davidjohnoliver) [inline doc/articles/uno-development/api-extensions.md]:
```suggestion
When a NuGet package containing such a declaration is found in the currently built application, the App.InitializeComponent() method will automatically add the following code to the app startup:
```

File: doc/articles/uno-development/api-extensions.md (hunk 1 above)
Comment (davidjohnoliver) [reply doc/articles/uno-development/api-extensions.md]: Seems great - will we able to use this for eg the map control?

File: src/Uno.Foundation/Extensibility/ApiExtensibility.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace Uno.Foundation.Extensibility
+{
+	/// <summary>
+	/// Registry for API existensibility providers, used to provide optional
+	/// implementations for compatible parts of WinUI and WinRT APIs.
+	/// </summary>
+	public static class ApiExtensibility
+	{
+		private static object _gate = new object();
```
Comment (davidjohnoliver) [inline src/Uno.Foundation/Extensibility/ApiExtensibility.cs]:
```suggestion
		private static readonly object _gate = new object();
```

File: src/Uno.Foundation/Uno.Foundation.csproj [hunk 5]
```diff
@@ -40,6 +40,11 @@
 		<PackageReference Include="System.Runtime.InteropServices.WindowsRuntime" Version="4.3.0" />
 	</ItemGroup>
 
+	<ItemGroup>
+	  <UpToDateCheckInput Remove="Extensibility\ApiExtensibility.cs" />
```
Comment (davidjohnoliver) [inline src/Uno.Foundation/Uno.Foundation.csproj]: Remove these changes

File: src/Uno.UI.DualScreen/DuoApplicationViewSpanningRect.cs [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+namespace Uno.UI.DualScreen
+{
+	public class DuoApplicationViewSpanningRects : IApplicationViewSpanningRects
+	{
+		Microsoft.Device.Display.ScreenHelper _helper;
+		List<Rect> _spanningRects = null;
+
+		private readonly List<Rect> EmptyList = new List<Rect>(0);
+
+		public DuoApplicationViewSpanningRects(object owner)
```
Comment (davidjohnoliver) [inline src/Uno.UI.DualScreen/DuoApplicationViewSpanningRect.cs]: Unused parameter?

File: src/Uno.UI/Uno.UI.csproj [hunk 7]
```diff
@@ -357,11 +357,21 @@
 		</ProjectReference>
 	</ItemGroup>
 
+	<ItemGroup>
+	  <UpToDateCheckInput Remove="Microsoft\UI\Xaml\Controls\TwoPaneView\DisplayRegionHelperTestApi.cs" />
```
Comment (davidjohnoliver) [inline src/Uno.UI/Uno.UI.csproj]: Remove these changes

File: src/Uno.UWP/UI/ViewManagement/ApplicationView.Android.cs [hunk 8]
```diff
@@ -1,5 +1,6 @@
 ﻿#if __ANDROID__
 using System;
+using System.Linq;
```
Comment (davidjohnoliver) [inline src/Uno.UWP/UI/ViewManagement/ApplicationView.Android.cs]: Unnecessary?

File: src/Uno.UWP/Uno.csproj [hunk 9]
```diff
@@ -38,6 +38,10 @@
 	<ItemGroup>
 		<UpToDateCheckInput Include="**\*.cs" Exclude="bin\**\*.cs;obj\**\*.cs;" />
 	</ItemGroup>
+	
+	<ItemGroup>
+	  <UpToDateCheckInput Remove="UI\ViewManagement\IApplicationViewSpanningRects.cs" />
```
Comment (davidjohnoliver) [inline src/Uno.UWP/Uno.csproj]: Remove

File: doc/articles/uno-development/api-extensions.md (hunk 1 above)
Comment (jeromelaban) [reply doc/articles/uno-development/api-extensions.md]: We could. There's a need to make sure that interfaces are available for the override, but it's the overall idea.

File: src/Uno.UI.DualScreen/DuoApplicationViewSpanningRect.cs (hunk 6 above)
Comment (jeromelaban) [reply src/Uno.UI.DualScreen/DuoApplicationViewSpanningRect.cs]: Yes, but it's based on what the code generation expects. At this point, the constructor has to have one parameter.

**Pull request merged**